// @splitsoftware/splitio-angular/esm2015/lib/utils/constants.js
/**
 * SDK constant for control treatment
 */
export const CONTROL = 'control';
/**
 * string constant for observable to return when client exists for a key
 */
export const INIT_CLIENT_EXISTS = 'init::clientExists';
/**
 * string constant for observable to return when client is not initialized
 */
export const INIT_CLIENT_FIRST = 'init::clientFirst';
export const VERSION = 'angular-' + '0.3.0';
const ɵ0 = () => { return CONTROL; }, ɵ1 = () => { return { treatment: CONTROL, config: null }; }, ɵ2 = (splitNames) => {
    let result = {};
    splitNames.forEach((splitName) => {
        result = Object.assign(Object.assign({}, result), { [splitName]: CONTROL });
    });
    return result;
}, ɵ3 = (splitNames) => {
    let result = {};
    splitNames.forEach((splitName) => {
        result = Object.assign(Object.assign({}, result), { [splitName]: { treatment: CONTROL, config: null } });
    });
    return result;
}, ɵ4 = () => { return false; };
/**
 * client with methods that return default values
 */
export const CONTROL_CLIENT = {
    getTreatment: ɵ0,
    getTreatmentWithConfig: ɵ1,
    getTreatments: ɵ2,
    getTreatmentsWithConfig: ɵ3,
    track: ɵ4
};
const ɵ5 = () => { return []; }, ɵ6 = () => { return null; }, ɵ7 = () => { return []; };
/**
 *  with methods that return default values
 */
export const DEFAULT_MANAGER = {
    splits: ɵ5,
    split: ɵ6,
    names: ɵ7
};
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7 };
